### ISSUE TITLE
Early return from an initializer leaves fields uninitialized 

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#18524]
Having an early return in an initializer leaves the fields uninitialized and does not produce a warning or error to indicate that. 

In the following example the early return leave the fields uninitialized and so their default values are just random stack memory:

```chpl
use CPtr;
record lowLevelLocalizingSlice {
    type t;
    var ptr: c_ptr(t) = c_nil;
    var isOwned: bool = false;
    var initSentinel = 314;

    proc init(A: [] ?t, region: range(?)) {
      this.t = t;
      if region.isEmpty() {
        return;
      }

      this.ptr = c_malloc(t, region.size);
      this.isOwned = true;
      for i in 0..<region.size {
        this.ptr[i] = A[region.low + i];
      }
    }

    proc deinit() {
      if isOwned then c_free(ptr);
    }
}

var A: [1..10] int;
writeln(new lowLevelLocalizingSlice(A, 1..1));
writeln(new lowLevelLocalizingSlice(A, 1..0));
```

```
(ptr = 0x7f8492604280, isOwned = true, initSentinel = 314)
(ptr = nil, isOwned = false, initSentinel = 1) // initSentinel should be 314
```

In this small test case, it's easy to change to `if !region.isEmpty() { //realInit }` or something, but the original code is from https://github.com/Bears-R-Us/arkouda/blob/be14087/src/AryUtil.chpl#L109-L159 where I find the early return makes the code easier to read compared to having extra branches and indentation in the "real" init code.

In an Arkouda test, the random stack memory we got caused `isOwned` to be true and so we tried to free a random address, which is what caused us to notice this issue.

I don't have a strong preference for inserting default field init before the return or just making this a user error, but something to prevent the error would be nice.

**Comment by mppf** [comment]
Wow, I thought that returning from an initializer was a full out compilation error.

**Comment by ronawho** [comment]
I'm not against making a return in an initializer illegal. I do find the early return code cleaner for exception cases, but it's not the worst thing to lose that.

**Comment by mppf** [comment]
I think this is related to making initializers throw (#8692) and they could both use a similar strategy that tracks the count of how many fields have been initialized at the time the function is being exited. (The difference is that a thrown error needs us to deinit any fields initialized so far; while a return would need us to default-init any remaining fields).

**Comment by bradcray** [comment]
I agree it's somewhat related (in that both are about taking exceptional paths through initializers), but given that Elliot's request was about having `return` imply a `this.complete();` it seems as though the current machinery that strives to find the place where a compiler-inferred `this.complete()` would be inserted, filling in any other field initializers, could potentially be leveraged without relying on an execution-time tracking mechanism (?).  (none of this is to say we shouldn't build the machinery that supports initializers that throw, however).

**Comment by lydia-duncan** [comment]
Elliot and I discussed this offline before he opened the issue, but just to make sure my position is known: my expectation is that leaving an initializer results in the fields being given an initial value.  I don't have a strong preference between whether that means this case is an error or whether we ensure the fields are initialized before leaving.  I do think leveraging the existing `this.complete()` logic is the right call in this case, because the expectation is that the object is in a good state when we leave it.  I would expect the handling to be different for thrown errors because the result is different
